Make EnsureAuthenticatedLinks compatible with AppBridge 2

kirillplatonov opened the pull request:
### What this PR does

`EnsureAuthenticatedLinks` is currently broken and doesn't pass `host` header for `AppBridge 2`. As a result, all deep links in the app are broken.
![CleanShot 2021-06-05 at 20 21 22@2x](https://user-images.githubusercontent.com/839922/120900808-a6e1fe80-c63f-11eb-8996-67449f41bc29.png)

I added the missing "host" header to redirect to the splash page to fix it.

### Checklist

Before submitting the PR, please consider if any of the following are needed:

- [x] Update `CHANGELOG.md` if the changes would impact users
- [x] Update `README.md`, if appropriate.
- [x] Update any relevant pages in `/docs`, if necessary
- [x] For security fixes, the [Disclosure Policy](https://github.com/Shopify/shopify_app/blob/master/SECURITY.md#disclosure-policy) must be followed.

rs-ehyde commented:
We are trying to get our application listed and we were just told that AppBridge 2.0 is a **requirement**. Pretty wild that this is still broken.

kirillplatonov commented:
@NabeelAhsen @rezaansyed @andyw8
Hey. App Bridge 2.0 is a hard requirement for new app submissions. Any chance this bugfix can be merged?

artyrcheek commented:
> We are trying to get our application listed and we were just told that AppBridge 2.0 is a **requirement**. Pretty wild that this is still broken.

Crazy this still isnt fixed

sshaw commented:
> We are trying to get our application listed and we were just told that AppBridge 2.0 is a requirement. Pretty wild that this is still broken.

This is why I avoid embedded apps. Too many headaches!

kirillplatonov commented:
@BranLiang I saw you merged another PR in this repo recently. Could you please review this one too?

NabeelAhsen reviewed (approved):
Hey it's been a while since I actively contributed to this gem. It seems that this PR keeps slipping through the cracks, I can help ship + release this change.

I'm approving this to unblock - but let's get this PR up to speed with the main branch and add some more test coverage in the event where `host` isn't specified as a param by Shopify.

kirillplatonov commented on app/controllers/concerns/shopify_app/ensure_authenticated_links.rb:
```diff
@@ -10,9 +10,26 @@ module EnsureAuthenticatedLinks
+        .merge(params)
+        .to_query
+      uri.to_s
+    end
```
I just pushed a simplified version of this. Since we have a predefined set of params here (`return_to`, `shop`, `host`) there's no need to parse the existing query.

kirillplatonov commented:
@NabeelAhsen Hey and Happy new year!
Thanks for taking the time and looking at this 🙏. I rebased PR from master and simplified changes a bit.

We have a few more PR's that needs some attention to fully support AppBridge 2, improve JWT integration, and make the engine compatible with Ruby 3 and Rails 7. We're using [this fork](https://github.com/kirillplatonov/shopify_app) for these changes but it would be awesome to move these changes to the official repo. If you don't mind I will update them and mention you so we could merge & ship the new release.

mjesar commented:
Hi 
I have upgraded to shopify_app to (18.0.3) because getting this error `[ShopifyApp::EnsureAuthenticatedLinks] Redirecting to login: [ShopifyApp::LoginProtection::ShopifyDomainNotFound] Could not determine current shop domain`

but after upgrading shopify_app to  (18.0.3), I am still getting the same error on activate charge and frame  reload in the browser 
![error charge](https://user-images.githubusercontent.com/27009137/149086771-15f1e16c-8bdc-42be-8ec5-5270258898b9.png)

Any help regarding this will be appreciated
